SCP Code Refactor

TheDarkElites opened the pull request:
## About the Pull Request

Massively refactors both general backend SCP code and individual SCP code.

Moves memetic-related features of scps to a generalized memetic component.

## Why It's Good For The Game

EWWWW ~~BAYCODE~~ LIONCODE EWWWW SHITCODE

## Changelog

:cl:
refactor: Refactors SCP Backend Code
balance: Memetic system min dark value changed to 0.03 from 0.05
refactor: Refactors SCP 012
refactor: Refactors SCP 013
add: Adds SCP 017 (For the most part, its not mapped in)
refactor: SCP 049 has been refactored and minorly reworked
code: New stageHandler to allow scps to keep track of various stages in humans without explicitly defined vars.
fix: desc not showing SCP desc for humanoid SCPs.
code: new memetics component for handling memetic like properties of SCPS (but it can be used for anything).
add: SCP 066 now has actual functionality
refactor: SCP 096 has been refactored and reworked but is currently disabled as it requires AI ports that are still being worked on.
refactor: SCP 106 has been minorly refactored to use new SCP datum and stage handler.
refactor: SCP 113 has been refactored.
refactor: SCP 131 has been refactored and slightly reworked.
refactor: SCP 151 has been refactored and reworked.
refactor: SCP 173 has been minorly refactored.
refactor: SCP 216 has been minorly refactored.
refactor: SCP 247 has been refactored.
refactor: SCP 263 majorly reworked.
add: SCP 280 is no longer a placeholder and now works
refactor: SCP 294 was minorly reworked to take reagents from the world rather than just being a chem dispenser.
refactor: SCP 343 has been refactored
refactor: SCP 420-J has been reworked (credit to @EgorDinamit)
refactor: SCP 457 has been refactored and minorly reworked
refactor: SCP 500 has been refactored
refactor: SCP 513 has been fully reworked.
refactor: SCP 527 minorly refactored.
refactor: SCP 529 minorly refactored.
refactor: SCP 895 reworked.
refactor: SCP 999 refactored.
refactor: SCP 1102RU refactored.
refactor: SCP 2343 refactored (just made a subtype of 343)
refactor: SCP 2398 minorly reworked.
refactor: SCP 2427-3 minorly refactored.
code: new is_dark() turf helper proc to help standardize what is considered dark.
add: new functionality to allow ladder types to be exempt from unit tests.
fix: gets rid of duplicate English language
del: gets rid of all placeholder SCP files
fix: should fix #1141
fix: should fix #1190
fix: fix #207
fix: SCP 2427-3 no longer checks purity of SCPs.
fix: fixes #1295
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

<!--
Examples were changelog entries are optional/not typically required but encouraged:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

As a courtesy, for ported PRs, please include if you have the blessing of the author. While not required, we encourage basic decency in porting others' work. It is also sufficient to note that the author has not expressed displeasure at the idea of their work getting ported.
Please refrain from porting works of authors that have expressed displeasure in having their work ported, thank you.
-->

TheDarkElites committed:
Initial SCP backend work and memetic component + 012 refactor

Tsurupeta commented:
its lioncode, not baycode

TheDarkElites committed:
checks

TheDarkElites commented:
> its lioncode, not baycode

fixed

TheDarkElites committed:
fixes ambiguity

TheDarkElites committed:
more ambiguity fixes

TheDarkElites committed:
013 Refactor + Stage Handler Datum

TheDarkElites committed:
013 smoke infinite until effects complete

also removes unneeded scp helpers

TheDarkElites committed:
blue lady fixes

TheDarkElites committed:
scp helpers are back

also changes datum to ref correct var

(all atoms have a fucking SCP datum var for some fucking reason)

TheDarkElites committed:
fixes helper error

TheDarkElites committed:
scp helper fixes

TheDarkElites committed:
fixes projectile.dm runtime

TheDarkElites committed:
SCP 017 added properly

TheDarkElites committed:
butchery values fix

TheDarkElites committed:
017 attackby and hitby fixes

TheDarkElites committed:
049 Refactor and rework, photography fixes, and humanoid examine fixes

TheDarkElites committed:
066 sound files

TheDarkElites committed:
AI holder sound bugfix

TheDarkElites committed:
Signal fixes

memetic integration with sound aswell

TheDarkElites committed:
faction fixes

TheDarkElites committed:
hologram runtime fix

TheDarkElites committed:
minor SCP fixes

TheDarkElites committed:
066 refactor and implementation

TheDarkElites committed:
placeholder flag + name standardization

TheDarkElites committed:
moves sounds to appropriate folders and comment dividers

TheDarkElites committed:
updates memetic component

TheDarkElites committed:
096 sound stuff

TheDarkElites committed:
096 Refactor

096 majorly refactored and reworked, however disabled until cheese ports appropriate AI systems.

TheDarkElites committed:
106 minor refactor

TheDarkElites committed:
improper fixes for SCP names

TheDarkElites committed:
runtime fix + name improper fixes

TheDarkElites committed:
minor code qol improvment

TheDarkElites committed:
human helpers and scp log minor update

TheDarkElites committed:
113 Refactor

TheDarkElites committed:
113 update

TheDarkElites committed:
uses apply damage instead of adjustBruteLoss in 012

TheDarkElites committed:
113 groin damage update

TheDarkElites committed:
fixes issues with convertatom2type not ghostizing when there is no mind to transfer to

TheDarkElites committed:
convertatom2type now returns new atom upon conversion

TheDarkElites committed:
convertatom2type updates and implementation

TheDarkElites committed:
minor convertatom2type proc fix

TheDarkElites committed:
scp datum faction assignment update

TheDarkElites committed:
131 Rework/Refactor + Eyepod language

TheDarkElites committed:
checks fix

TheDarkElites committed:
empty_stomach -> vomit on 113

TheDarkElites committed:
151 refactor/rework

TheDarkElites committed:
173 minor refactor

just moves us over to the new datum, nothing fancy

TheDarkElites committed:
178 refactor + 173 fix

TheDarkElites committed:
removes unneeded var/

TheDarkElites committed:
removes "_" from 173's path

TheDarkElites committed:
216 minor refactor + small rework

216 now generates items upon code entered, not at initialization. Also more config options now.

TheDarkElites committed:
fixes checks

TheDarkElites committed:
247 refactor

TheDarkElites committed:
updates signal comments to follow format

TheDarkElites committed:
language fix for eyepods

TheDarkElites committed:
263 refactor

TheDarkElites committed:
adds desg to placeholder message

TheDarkElites committed:
fixes 247 intent check

TheDarkElites committed:
moves 078 to correct folder

TheDarkElites committed:
078 set as placeholder

it was poorly implemented as is and it would be too difficult to implement fully now.

TheDarkElites committed:
initialize and barsign fixes

TheDarkElites committed:
scp updates

TheDarkElites committed:
scp 294 rework

TheDarkElites committed:
343 refactor

TheDarkElites committed:
263 fixes

TheDarkElites committed:
280 initial implementation

TheDarkElites committed:
399 and 420

EgorDinamit commented on code/_helpers/_scp.dm:
```diff
@@ -0,0 +1,7 @@
+//SCP Helpers
+
+///Checks if atom is an SCP
+/proc/isSCP(atom/A)
```
This seems a bit redundant, no?
Would make sense if it was an atom proc, so you could override results on case-by-case basis, but as it is right now - it is rather pointless and you can simply check if(A.SCP) instead.

TheDarkElites replied:
Yes, this was when I just started, I just ported all existing stuff and decided to wait until I do the final cleanup and pass over stuff to see if it's necessary. Probably will remove it before I mark this as ready for review.

EgorDinamit commented on code/_helpers/turfs.dm:
```diff
@@ -112,6 +112,11 @@
 	return FALSE
 
+/proc/is_dark(turf/T, darkness_threshold = 0.3)
+	if(T.get_lumcount() <= 0.3)
```
```suggestion
	if(T.get_lumcount() <= darkness_threshold)
```
Pretty sure you meant to do this, judging by this particular unused proc argument.

TheDarkElites replied:
Lmao I cant believe I missed that

TheDarkElites committed:
minor rework for 420 (Mostly stolen from @EgorDinamit 's 914 PR)

TheDarkElites committed:
Update code/_helpers/turfs.dm

Co-authored-by: EgorDinamit <53223414+EgorDinamit@users.noreply.github.com>

TheDarkElites committed:
fixes checks and moves over to JPS pathfinding

TheDarkElites committed:
minor fixes

TheDarkElites committed:
457 initial implementation

TheDarkElites commented:
this is so fucking painful

TheDarkElites committed:
457 playable fix

TheDarkElites committed:
scp 500 refactor

TheDarkElites committed:
113 icon fixes

TheDarkElites committed:
513 refactor/rework

TheDarkElites committed:
513 fix

TheDarkElites committed:
527 minor refactor

TheDarkElites committed:
529 minor refactor

TheDarkElites committed:
714

TheDarkElites committed:
comments out scp verb

TheDarkElites committed:
Memetics now handle camera view

TheDarkElites committed:
895 refactor

TheDarkElites committed:
deletes placeholder file for 966

theres just some shitty code and not even a icon so I see no point in keeping a placeholder for this.

TheDarkElites committed:
finishes removing 966 file

TheDarkElites committed:
999 minor refactor

TheDarkElites committed:
999 minor code organization

TheDarkElites committed:
1102ru Refactor

TheDarkElites committed:
1499

TheDarkElites committed:
2343

TheDarkElites committed:
scp 2398 refactor

TheDarkElites committed:
adds swing time config for 2398

TheDarkElites committed:
2427-3 and 2343 minor refactor

TheDarkElites committed:
gets rid of uneeded 106 vars

TheDarkElites committed:
3349 minor refactor

TheDarkElites committed:
updates SCP possession verb and handling to not be trash

TheDarkElites committed:
implementation of new is_dark helper where necessary

TheDarkElites committed:
checks fix maybe?

TheDarkElites committed:
checks fix for real this time

TheDarkElites committed:
more checks fix

TheDarkElites committed:
012 second pass over

TheDarkElites committed:
1102 second pass over

hopefully this also passes checks

TheDarkElites committed:
heart runtime fix

TheDarkElites committed:
heart fix please I beg

TheDarkElites committed:
please heart fix

TheDarkElites committed:
heart fix four, the heartening

TheDarkElites committed:
more 1102 fixes

TheDarkElites committed:
adds the ability to exclude ladder types to ladder unit tests

guh

TheDarkElites committed:
012 audio fix

TheDarkElites committed:
013 second pass over

TheDarkElites committed:
gets rid of pick(list())

pick(list()) -> pick()

TheDarkElites committed:
049-1 second pass over + plaguespeak

TheDarkElites committed:
fixes 049's door handling

TheDarkElites committed:
language fixes

TheDarkElites committed:
049 regen mechanics and further fixes

TheDarkElites committed:
checks fix

TheDarkElites committed:
final 049 passover

TheDarkElites committed:
code fixes

TheDarkElites committed:
Deletes placeholder files

How the fuck this even got passed I dont know but its dumb and stupid and I hate it.

TheDarkElites committed:
066 second passover

TheDarkElites committed:
SCP_PLACEHOLDER flag replaced with SCP_DISABLED

TheDarkElites committed:
should fix #1190

TheDarkElites committed:
sound fix

TheDarkElites committed:
fixes #207

TheDarkElites committed:
sound fix pt.2

TheDarkElites committed:
066 autohiss

TheDarkElites committed:
096 second pass over

TheDarkElites committed:
106 config organization

TheDarkElites committed:
moves 008 into its own file

TheDarkElites committed:
updates 049 instances to use updated count

TheDarkElites committed:
checks fix

TheDarkElites committed:
more checks fix

TheDarkElites committed:
008 fix and 2427-3 fix

TheDarkElites committed:
113 second pass over

TheDarkElites committed:
131 second passover

TheDarkElites committed:
fixes threaten runtime

TheDarkElites committed:
minor 106 refactor

TheDarkElites committed:
247 second pass over

TheDarkElites committed:
263 second passover

TheDarkElites committed:
294 second passover

TheDarkElites committed:
173 status tab now shows amount of time before breach

TheDarkElites committed:
standardizes names for 2427-3

TheDarkElites committed:
further name standardization (216)

TheDarkElites committed:
macro fix

TheDarkElites committed:
999 minor fix

TheDarkElites committed:
name standardization

TheDarkElites committed:
name fixes

TheDarkElites committed:
final name fix

TheDarkElites committed:
checks fix

TheDarkElites committed:
removal of isSCP function

TheDarkElites committed:
280 finalizes

TheDarkElites committed:
tgui span styling fixes + scp desg mechanics

TheDarkElites committed:
fixes runtimes

TheDarkElites committed:
whoops

TheDarkElites committed:
fixes 106 not doing damage in his dimension and 457 respawn issues

TheDarkElites committed:
adds checks for pocket dimension verbs

TheDarkElites committed:
whoops

TheDarkElites committed:
runtime fix

TheDarkElites committed:
typo fix + 457 improvment

TheDarkElites committed:
runtime fix

TheDarkElites committed:
minor fix for 096

TheDarkElites committed:
runtime fixes

TheDarkElites committed:
updates cage mechanics

173 should be able to damage cage while moving, it should only be dependent on if it is being watched.

TheDarkElites committed:
runtime fix for 2427-3

TheDarkElites committed:
dead scps no longer show up as possesible

Tsurupeta commented on code/__defines/SCP.dm:
```diff
@@ -1,9 +1,35 @@
-#define SCP_096 "096"
+// Classifcations
 
+#define SAFE		"Safe"
+#define EUCLID		"Euclid"
+#define KETER 		"Keter"
+#define THAUMIEL	"Thaumiel"
+#define NEUTRALIZED "Neutralized"
```
Add SCP_ before the define

TheDarkElites committed:
Makes define names SCP_ related

Tsurupeta commented on code/_helpers/type2type.dm:
```diff
@@ -215,3 +216,38 @@
+	var/atom_loc = get_turf(target_atom)
+	var/atom/new_atom = new type(atom_loc)
+
+	if(ismob(target_atom))
+		var/mob/target_mob = target_atom
+		var/mob/new_mob = new_atom
+
+		for(var/obj/item/W in target_mob)
+			target_mob.drop_from_inventory(W)
+		target_mob.regenerate_icons()
+
+		if(!target_mob.mind)
+			qdel(target_atom)
+			return new_atom
+		if(convert_message)
+			to_chat(target_mob, convert_message)
+		if(!istype(new_mob) || (!new_mob.mind && !force_mind))
+			target_mob.ghostize()
+			qdel(target_atom)
+			return new_atom
```
Why regenerate icons on mob that's going to be deleted anyways and also why use target_atom for qdel if you already have target_mob local var?

Tsurupeta commented on code/datums/components/memetic.dm:
```diff
@@ -0,0 +1,119 @@
+	var/list/memetic_sounds
+
+/datum/component/memetic/Initialize(flags, meme_proc, memeticSounds)
+	.=..()
```
spaces dude

Tsurupeta commented on code/datums/mob_stagehandler.dm:
```diff
@@ -0,0 +1,44 @@
+
+///Creates a new stage to track
+/datum/stageHandler/proc/createStage(stageID)
+	if(stages[stageID])
```
It won't trigger on stage being equal to 0, so you need to check for null. Intended?

Tsurupeta commented on code/datums/mob_stagehandler.dm:
```diff
@@ -0,0 +1,44 @@
+
+///Returns value of a stage
+/datum/stageHandler/proc/getStage(stageID)
+	if(!stages[stageID])
```
This also checks for 0, not null, so after creation it gets set to 0 but gets treated as not existing

Tsurupeta commented on code/game/atoms.dm:
```diff
@@ -282,6 +282,11 @@ its easier to just keep the beam vertical.
 	to_chat(user, desc)
 	if(health_max)
 		examine_damage_state(user)
+	if(SCP && ishuman(user))
+		var/mob/living/carbon/human/H = user
+		var/datum/job/job = SSjobs.get_by_title(H.job)
+		if(job && (job.department_flag & (COM|SCI|SEC)))
+			to_chat(user, SPAN_CLASS("scp", "You know this is SCP-[SCP.designation]!"))
```
Maybe we can move it out somewhere else????

TheDarkElites replied:
might be possible to move it to the SCP component, will check

77seven77seven replied:
ghosts being able to know what SCP it is would be good too

Tsurupeta commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -372,8 +372,9 @@
-	heart.scp3349_induced = FALSE	// part of SCP-3349. electric shocks temporarily stop effect
-	to_chat(H, SPAN_INFO("The euphoric sensation ends."))
+	if(heart.SCP.designation == "3349-1")	// part of SCP-3349. electric shocks temporarily stop effect
+		SCP.Destroy()
```
????????

TheDarkElites replied:
This was a way of checking for the SCP instance existing without using a specialized var in the heart organ.

Tsurupeta replied:
Why is Destroy() called here? It's not meant to be called directly like that.

TheDarkElites replied:
Err I think at the time I wasn't sure how to handle nulling a datum and the datum's destroy proc handles getting everything done neatly

Lmk how it should be done

Tsurupeta reviewed (changes requested):
Huh

Tsurupeta commented on code/datums/components/memetic.dm:
```diff
@@ -0,0 +1,119 @@
+/datum/component/memetic/proc/activate_memetic_effects()
+	if(!(memetic_flags & MPERSISTENT)) //if we arent a persistent memetic, then affected humans will be removed from effect after they no longer meet the reqs
+
+		for(var/mob/living/carbon/human/H in affected_mobs)
```
wrong usage of weakrefs

Tsurupeta replied:
I think I explained how t use them properly, I forgot...

Tsurupeta commented on code/game/objects/items/devices/scanners/health.dm:
```diff
@@ -117,7 +117,7 @@
 	var/pulse_result = "normal"
 	var/obj/item/organ/internal/heart/heart = H.internal_organs_by_name[BP_HEART]
-	if(H.should_have_organ(BP_HEART) && !(heart.scp3349_induced))
+	if(H.should_have_organ(BP_HEART) && (heart.SCP.designation != "3349-1"))
```
```suggestion
	if(H.should_have_organ(BP_HEART) && (heart.SCP?.designation != "3349-1"))
```
???

Tsurupeta commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -372,8 +372,9 @@
 	var/obj/item/organ/internal/heart/heart = H.internal_organs_by_name[BP_HEART]
-	heart.scp3349_induced = FALSE	// part of SCP-3349. electric shocks temporarily stop effect
-	to_chat(H, SPAN_INFO("The euphoric sensation ends."))
+	if(heart.SCP.designation == "3349-1")	// part of SCP-3349. electric shocks temporarily stop effect
```
```suggestion
	if(heart.SCP?.designation == "3349-1")	// part of SCP-3349. electric shocks temporarily stop effect
```
???

Tsurupeta commented on code/modules/SCP/SCPs/SCP-013.dm:
```diff
@@ -1,26 +1,113 @@
+	if(lit)
+		if(H.humanStageHandler.createStage("BlueLady"))
+			update_013_status(H)
+			affected += H
```
wrong weakref usage

Tsurupeta commented on code/modules/SCP/SCPs/SCP-013.dm:
```diff
@@ -1,26 +1,113 @@
+			addtimer(CALLBACK(src, .proc/update_013_status, H), 55 SECONDS)
+		if(7)
+			addtimer(CALLBACK(H, /mob/living/carbon/human/proc/bluelady_message), 10 SECONDS)
+			affected -= H
```
wrong weakref usage

EgorDinamit commented:
God I wish this PR didn't have that many files changed...

TheDarkElites commented:
> God I wish this PR didn't have that many files changed...

If you are asking me to do less work I can do that too

TheDarkElites committed:
nullchecks pt1

Co-authored-by: Tsurupeta <41485301+Tsurupeta@users.noreply.github.com>

TheDarkElites committed:
nullchecks pt2

Co-authored-by: Tsurupeta <41485301+Tsurupeta@users.noreply.github.com>

TheDarkElites committed:
fixes

TheDarkElites committed:
moves examine message to scp datum

TheDarkElites committed:
fixes weakrefs

TheDarkElites committed:
Makes 343 an advanced tool user again

TheDarkElites committed:
unlocks 066 door

TheDarkElites committed:
makes 457 immune to fire damage

how did we not think of this

TheDarkElites committed:
173 breakout cooldown does not activate cooldown if activated while being watched

EgorDinamit commented on code/modules/SCP/scp.dm:
```diff
@@ -1,78 +1,112 @@
 /datum/scp/Destroy()
 	. = ..()
-	if(GLOB.SCP_list.len)
+	if(LAZYLEN(GLOB.SCP_list))
 		GLOB.SCP_list -= src
```
The entire length check is pointless, as `-=` won't delete anything if it isn't in the list.

EgorDinamit commented on code/modules/SCP/scp.dm:
```diff
@@ -1,78 +1,112 @@
 		qdel(src)
 
+///For when an SCP object is examined, we send the examinee a message about the SCP's designation if they should know what SCP it is.
+/datum/scp/proc/OnExamine(datum/source, mob/examinee)
+	SIGNAL_HANDLER
+	if(!ishuman(examinee))
+		return
+	var/mob/living/carbon/human/H = examinee
+	var/datum/job/job = SSjobs.get_by_title(H.job)
+	if(job && (job.department_flag & (COM|SCI|SEC)))
+		to_chat(examinee, SPAN_CLASS("scp", "You know this is SCP-[designation]!"))
```
Should probably be available per some trait instead of jobs.

TheDarkElites replied:
Once TG traits are ported then sure but for the time being we don't really have an all-encompassing trait system.

EgorDinamit commented on code/modules/mob/language/human/english.dm:
```diff
@@ -1,21 +1,27 @@
+		LANGUAGE_HUMAN_INDIAN = 5,
+		LANGUAGE_HUMAN_SPANISH = 15,
+		LANGUAGE_HUMAN_JAPANESE = 5,
+		LANGUAGE_HUMAN_RUSSIAN = 5,,
```
Random double comma.

EgorDinamit commented on code/modules/SCP/SCPs/SCP-294.dm:
```diff
@@ -1,104 +1,137 @@
-		"venom" = /datum/reagent/toxin/venom,
 		"spider" = /datum/reagent/toxin/venom,
-		// Sedatives
 		"sleep" = /datum/reagent/soporific,
-		"chloral" = /datum/reagent/chloral_hydrate,
-		"zombie" = /datum/reagent/toxin/zombie_powder,
+		"kidnap" = /datum/reagent/chloral_hydrate,
+		"zombie" = /datum/reagent/scp008,
+		"death" = /datum/reagent/toxin/cyanide,
+		"joe" = /datum/reagent/blood,
+		"008" = /datum/reagent/scp008,
+		"health" = /datum/reagent/scp500
```
You removed a huge chunk of shortcut names. Any reason for that?

TheDarkElites replied:
Medicine shortcuts were not conceptual shortcuts like the words medicine, knowledge, poison, etc. but instead literal shortcuts. That doesn't really make sense for 294.

EgorDinamit commented on code/modules/SCP/SCPs/SCP-294.dm:
```diff
@@ -1,104 +1,137 @@
+	)
+//Mechanics
+
+///Cycles through all reagents datums and picks out ones that contain the chemical we are looking for
```
This may break a lot of things, leading to empty food/pills/medipens.
I am honestly not sold on this idea, as you also removed the blacklisted chems, making taking SCP500 and 008 incredibly easy.

TheDarkElites replied:
294 will have to be moved to a more restricted area, like LCZ.

Also, I will have interactions with containers that 294 is transferring from have a visual effect + some sort of message.

Ill also probably have pills, injectors, and other one-use chem containers delete themselves if they become empty.

TheDarkElites replied:
I'll do this in a PR right after this one as I dont want delay this being merged any further

TheDarkElites committed:
disables 343 - requested by Windbomb

TheDarkElites committed:
implements 457 burn resistance properly

TheDarkElites committed:
fixes reviews

TheDarkElites committed:
updates 294 to only poll playable z-levels

TheDarkElites committed:
Fixes 294 not polling reagents in containers

Tsurupeta commented on code/_helpers/lists.dm:
```diff
@@ -779,3 +779,26 @@ Checks if a list has the same entries and values as an element of big.
+	return output_list
+
+///Removes a weakref from a list which references the passed ref.
+/proc/removeWeakrefFromList(list/target_list, reference)
+	for(var/item in target_list)
+		if(isweakref(item))
+			var/weakref/weakrefitem = item
+			if(weakrefitem.resolve() == reference)
+				target_list -= item
+		else
+			continue
```
idk

Tsurupeta replied:
ok wait thats a stupid idea

Tsurupeta replied:
Just get ref of object and remove it from list...

TheDarkElites committed:
implements do_after PR

TheDarkElites committed:
fixes errors

TheDarkElites committed:
reamoves removeWeakrefFromList proc

TheDarkElites committed:
adds blacklist ability to 294

This is a temporary solution until 294 is moved to a more secure area

TheDarkElites committed:
ambiguity fix

Tsurupeta commented on code/datums/components/memetic.dm:
```diff
@@ -0,0 +1,126 @@
+
+		for(var/weakref/Hweakref in affected_mobs_weakref)
+			var/mob/living/carbon/human/H = Hweakref.resolve()
+			if((memetic_flags & MVISUAL) && H.can_see(parent, TRUE))
```
check for istype H on all weakref resolves because it can return bad thing

Tsurupeta commented on code/_helpers/lists.dm:
```diff
@@ -779,3 +779,16 @@ Checks if a list has the same entries and values as an element of big.
+//The following deal with lists of weakrefs
+
+///Takes a list of weakrefs and returns a list containing all the resolved weakrefs.
+/proc/resolveWeakrefList(list/input_list)
+	var/list/output_list
+	for(var/item in input_list)
+		if(isweakref(item))
+			var/weakref/weakrefitem = item
+			output_list += weakrefitem.resolve()
+		else
+			output_list += item
+	return output_list
```
```suggestion
/proc/resolveWeakrefList(list/input_list)
	var/list/output_list
	for(var/item in input_list)
		if(isweakref(item))
			var/weakref/weakrefitem = item
			var/datum/D = weakrefitem.resolve()
			if(D)
				output_list += D
		else
			output_list += item
	return output_list
```

Tsurupeta commented on code/datums/components/memetic.dm:
```diff
@@ -0,0 +1,126 @@
+/datum/component/memetic/proc/activate_memetic_effects()
+	if(!(memetic_flags & MPERSISTENT)) //if we arent a persistent memetic, then affected humans will be removed from effect after they no longer meet the reqs
+
+		for(var/weakref/Hweakref in affected_mobs_weakref)
```
also you have the resolve weakrefs list proc, why are you doing it manually here?

TheDarkElites replied:
Because I am a foolish man bound by the constraints of my material form

(I made the resolve proc after writing this and forgor, will fix)

Tsurupeta commented on code/modules/SCP/scp.dm:
```diff
@@ -1,78 +1,111 @@
+/datum/scp/New(atom/creation, vName, vClass = SCP_SAFE, vDesg, vMetaFlags)
+	GLOB.SCP_list += creation
+
+	name = vName //names are now usually captalized improper descriptors to fit the theme of SCP since people dont just know the scp desg off the bat. As such we need to improper it. TODO: add mental mechanic for foundation workers to see desg instead of name.
```
src.name = name

you can assign vars like this instead of doing vname and stuff, just for the future i guess

TheDarkElites replied:
good to know for future reference

Tsurupeta commented on code/modules/mob/living/carbon/hallucinations.dm:
```diff
@@ -233,11 +253,8 @@
-		I.dir = pick(GLOB.alldirs)
-		I.pixel_x = rand(-10,10)
-		I.pixel_y = rand(-10,10)
+		var/image/I = image('icons/effects/blood.dmi', "gib[pick("arm", "leg", "torso")]_flesh", layer = BELOW_TABLE_LAYER)
+		I.color = WOOD_COLOR_PALE
```
planes + layers for images are important

TheDarkElites replied:
layer is set in constructor :/

Tsurupeta commented on code/modules/organs/internal/heart.dm:
```diff
@@ -102,7 +100,7 @@
 
 	// So does SCP-3349
-	if(pulse != PULSE_NORM && scp3349_induced)
+	if(pulse != PULSE_NORM && ((SCP ? SCP.designation : "") == "3349-1"))
```
why ternary here?

TheDarkElites replied:
While all hearts have the SCP component, it is only initialized for hearts that are affected by 3349 (i.e., the hearts themselves becoming SCP instances), meaning that without this check, the code will throw runtimes.

Tsurupeta commented on code/modules/organs/internal/heart.dm:
```diff
@@ -117,9 +115,9 @@
 			heartbeat++
 
-		if(heartbeat >= (scp3349_induced ? (rate * 2) : rate))	// scp3349 heartbeat is long so we play it half as often to prevent overlap
+		if(heartbeat >= (((SCP ? SCP.designation : "") == "3349-1") ? (rate * 2) : rate))	// scp3349 heartbeat is long so we play it half as often to prevent overlap
```
and here

TheDarkElites replied:
See previous comment

Tsurupeta commented on code/modules/organs/internal/heart.dm:
```diff
@@ -117,9 +115,9 @@
+		if(heartbeat >= (((SCP ? SCP.designation : "") == "3349-1") ? (rate * 2) : rate))	// scp3349 heartbeat is long so we play it half as often to prevent overlap
 			heartbeat = 0
-			sound_to(owner, sound((scp3349_induced ? 'sounds/effects/heartbeatpurr.ogg' : beat_sound),0,0,0,50))
+			sound_to(owner, sound((((SCP ? SCP.designation : "") == "3349-1") ? 'sounds/effects/heartbeatpurr.ogg' : beat_sound),0,0,0,50))
```
and here

TheDarkElites replied:
See previous comment

Tsurupeta commented on code/modules/organs/internal/heart.dm:
```diff
@@ -235,7 +233,7 @@
 			. += "extremely fast and faint "
 
-	if(scp3349_induced)
+	if(SCP.designation == "3349-1")
```
```suggestion
	if(SCP?.designation == "3349-1")
```
??

Tsurupeta commented on code/modules/paperwork/photography.dm:
```diff
@@ -264,19 +254,24 @@ var/global/photo_count = 0
 	var/y_c = target.y + (size-1)/2
 	var/z_c	= target.z
 	var/mobs = ""
+	var/list/meta_data
+	LAZYINITLIST(meta_data)
```
```suggestion
	var/list/meta_data = list()
```
???

Tsurupeta commented on code/modules/paperwork/photography.dm:
```diff
@@ -31,16 +31,22 @@ var/global/photo_count = 0
-	var/anomalytype //This has to inherit from the camera
-	var/anomalymob //This has to inherit from the camera
+	///Photo data containing mobs and objects.
+	var/list/meta_data
```
isn't it holding refs?

TheDarkElites replied:
fixed to use weakrefs (and updated description)

Tsurupeta commented on code/modules/reagents/Chemistry-Holder.dm:
```diff
@@ -7,19 +7,22 @@ GLOBAL_DATUM_INIT(temp_reagents_holder, /obj, new)
 		build_chemical_reagent_list()
 	src.my_atom = my_atom
 	src.maximum_volume = maximum_volume
+	GLOB.reagents_datums += src
```
What is this list for?

EgorDinamit replied:
It seems to be used in 294's new way of finding the reagent holders.

EgorDinamit commented on code/__defines/misc.dm:
```diff
@@ -51,6 +51,7 @@
 #define STATUS_HUD_OOC 	9 // STATUS_HUD without virus DB check for someone being ill.
 #define LIFE_HUD 		10 // STATUS_HUD that only reports dead or alive
 #define BLINK_HUD       11 // A line representing the blink time remaining on a mob.
+#define PESTILENCE_HUD  12 // A visual indicator of the pestilence.
```
Clarify in the comment that it is specifically for 049's "pestilence". I am planning to add the virology eventually and it could be slightly confusing.

TheDarkElites committed:
Update code/modules/paperwork/photography.dm

Co-authored-by: Tsurupeta <41485301+Tsurupeta@users.noreply.github.com>

TheDarkElites committed:
Update code/modules/organs/internal/heart.dm

Co-authored-by: Tsurupeta <41485301+Tsurupeta@users.noreply.github.com>

TheDarkElites committed:
makes resolveWeakrefList safer

Co-authored-by: Tsurupeta <41485301+Tsurupeta@users.noreply.github.com>

TheDarkElites committed:
updates hud description

TheDarkElites committed:
updates memetic weakrefs to use new proc when enacting memetic effects

TheDarkElites committed:
updates photography meta data to hold weakrefs (and updates description)

TheDarkElites committed:
updates hallucination carnage to vary body parts

TheDarkElites commented:
ready for re-review or merge

Tsurupeta reviewed (approved):
Fuck it we ball

EgorDinamit committed:
Fix sound

EgorDinamit commented:
If it breaks - it breaks.
